File: Data Structures & Algorithms/search-2d-matrix/submission-1.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        int n = matrix[0].length;
        int low = 0;
        int high = m * n - 1;
        while(low <= high){
            int mid = (low + high) / 2;
            int xCoor = mid / n;
            int yCoor = mid % n;
            // System.out.println("low: "+low);
            // System.out.println("high: "+high);
            if(matrix[xCoor][yCoor] == target)
                return true;
            else if(target > matrix[xCoor][yCoor]){
                low = mid + 1;
            }else{
                high = mid - 1;
            }    
        }
        return false;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        low = 1
        high = max(piles)
        ans = high + 1
        while low <= high:
            mid = (low + high) // 2
            if self.canEat(mid, piles, h):
                ans = min(ans, mid)
                high = mid - 1
            else:
                low = mid + 1

        return ans        




    def canEat(self, speed: int, piles: List[int], h: int) -> bool:
        timeTaken = 0;
        for pile in piles:
            if pile % speed == 0:
                timeTaken += (pile // speed)
            else:
                timeTaken += ((pile // speed) + 1)        
        
        if timeTaken <= h:
            return True
        else:
            return False    